Feature/sign in refactor

aerych opened the pull request:
Closes #4912

This PR refactors the existing `LoginViewController` and implements a new "magic-links" sign-in feature.  The new controllers are written in Swift, and their UI is mainly defined in storyboard. View controllers are responsible for a single form, reducing the amount of state that needs to be maintained and tested.  Common functionality has been factored out into helper classes and protocols as much as possible, leaving controllers with a common, informal interface but no (or very limited) duplication. 


To test:
The `LoginViewController` supported a variety of sign-in scenarios, so there is a lot to test.  For each of the following scenarios, test the primary use-case, optional features on the form (various buttons) and error states (offline, invalid form entry). 

Scenario 1: 
Sign in to wpcom with an email address using the magic link flow.  To test the final step, reopening the app passing the auth token, you'll need to copy the `wordpress://` link from safari and change it to `wpdebug://`. 

Scenario 2: 
Sign in to wpcom with a username and password. No 2fa. 
 
Scenario 3: 
Sign in to wpcom with a username and password with 2fa.

Scenario 4: 
Sign in via a self-hosted blog. 

Scenario 5:
Confirm 1Password integration by testing on a device with 1Password installed.  Confirm that the 1Password button is present on the email, username/password, and self-hosted forms and that it functions on each. 
Note that on the self-hosted form you should be prompted to enter an URL if that field is empty. 

Scenario 6:
Confirm Safari stored password integration. 
Store credentials to wpcom 
- Settings > Safari > AutoFill > Names and Passwords.  
- Settings > Safari > Passwords > Add Password for wordpress.com
Launch and confirm you're prompted, and asked to use stored credentials. 
Choose credentials and confirm that they are filled out and you end up on the username/password form. 
Try again, and dismiss the prompt by clicking "Not Now".  Then re-open the prompt by clicking the saved passwords button on the Email form (first screen).


Needs review: @koke

koke commented on WordPress/Classes/Extensions/UITextField+TextHelper.swift:
```diff
@@ -0,0 +1,11 @@
+            return ""
+        }
+        return str
+    }
```
Interesting, I recently added `String.nonEmptyString` which is kind of the opposite approach: turn any empty string into a `nil` one. I find it a bit better to use optionals instead of empty strings as the compiler forces you to deal with the `nil` case in every occasion, but it's easy to forget handling an empty string.

aerych replied:
I can see that, tho I think I prefer to avoid optionals where ever possible.  Its a nice helper method to reduce the amount of optional unwrapping that would otherwise be required.

koke replied:
On a second thought I think I agree with you. I'm not sure why `UITextField.text` is an optional, and adding to the problem, the header says the default is `nil` and the documentation says the default is `""`. It's probably good to coalesce the empty cases to "".

I'm not sure if I'd add trimming to the mix, but the function name is clear enough about it :+1:

koke commented on WordPress/Classes/Networking/AccountServiceRemoteREST.m:
```diff
@@ -113,6 +117,72 @@ - (void)updateBlogsVisibility:(NSDictionary *)blogs
+- (void)findExistingAccountByEmail:(NSString *)email success:(void (^)(BOOL found))success failure:(void (^)(NSError *error))failure
+{
+    // TODO: We need to make a versioned flavor of this endpoint and ensure it always returns a JSON object.
+    // TODO: Remove the special case in `WordPressComApi.assertApiVersion` the endpoint is versioned.
```
We usually sign our `TODO`s with a name/date. Bonus points for a link to Github issue :wink:

koke commented on WordPress/Classes/Networking/AccountServiceRemoteREST.m:
```diff
@@ -113,6 +117,72 @@ - (void)updateBlogsVisibility:(NSDictionary *)blogs
+                  success(found);
+              }
+          } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
+              // The is-available endpoint is an oddball.
```
Definitely. Until there's a newer version of this API, I'd add a comment explaining what the possible responses are. This code made no sense until I did a few test requests to that endpoint and saw the responses myself.

aerych replied:
:+1: Comments added

koke replied:
I've just realized that the API isn't as odd as I thought (aside from the non-JSON response), but we're negating the semantics here. What if the remote method matched the API more closely:

```objc
- (void)isEmailAvailable:(NSString *)email success:(void (^)(BOOL found))success failure:(void (^)(NSError *error))failure;
```

Then the service's `findExistingAccountByEmail` would call this and negate the response.

Do you think that'd make this easier to follow?

aerych replied:
Maybe.  Let's try it.

kurzee commented:
Real quick, currently seeing a crash:
1. Sign in with a WordPress.com account.
2. On the My Sites tab > Switch Site > Hit the + to add a new site.
3. Select Add self-hosted site
4. Crash

```
'NSInvalidArgumentException', reason: 'NSConcreteAttributedString initWithString:: nil value'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010f0c7d85 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x000000010eb3bdeb objc_exception_throw + 48
	2   CoreFoundation                      0x000000010f0c7cbd +[NSException raise:format:] + 205
	3   Foundation                          0x0000000109b1d18b -[NSConcreteAttributedString initWithString:] + 140
	4   Foundation                          0x0000000109b1d051 -[NSConcreteAttributedString initWithString:attributes:] + 27
	5   WordPress                           0x00000001077e1c4a -[WPWalkthroughTextField commonInit] + 1370
	6   WordPress                           0x00000001077e1432 -[WPWalkthroughTextField init] + 114
	7   WordPress                           0x00000001077e14a2 -[WPWalkthroughTextField initWithLeftViewImage:] + 66
	8   WordPress                           0x00000001077d4cbf -[LoginViewController addControls] + 2287
	9   WordPress                           0x00000001077d11c4 -[LoginViewController viewDidLoad] + 420
	10  UIKit                               0x000000010afad984 -[UIViewController loadViewIfRequired] + 1198
```

aerych commented:
Nice catch @kurzee :)

koke commented on WordPress/Classes/Networking/ServiceRemoteREST.h:
```diff
@@ -4,6 +4,7 @@ typedef enum
 {
     ServiceRemoteRESTApiVersion_1_1 = 1001,
     ServiceRemoteRESTApiVersion_1_2 = 1002,
+    ServiceRemoteRESTApiVersion_1_3 = 1003,
```
Either I'm missing something or this doesn't seem to be used.

aerych replied:
Oops. We ended up using v1 of the REST API for the endpoint, not v3.  Old code. Will purge.

koke commented on WordPress/WordPressApi/WordPressComApi.m:
```diff
@@ -158,7 +158,8 @@ - (void)assertApiVersion:(NSString *)URLString
              || [URLString rangeOfString:@"v1.1"].length > 0
-             || [URLString rangeOfString:@"v1.2"].length > 0,
+             || [URLString rangeOfString:@"v1.2"].length > 0
+             || [URLString rangeOfString:@"is-available/email"].length > 0,
```
If the v1.3 is actually used, it should be added here as well.

aerych replied:
Nuked with fervor

koke commented on WordPress/Classes/ViewRelated/NUX/LoginFields.h:
```diff
@@ -11,6 +11,8 @@
 @property (nonatomic, assign) BOOL userIsDotCom;
 @property (nonatomic, assign) BOOL shouldDisplayMultifactor;
+@property (nonatomic, assign) NSInteger safariStoredUsernameHash;
+@property (nonatomic, assign) NSInteger safariStoredPasswordHash;
```
Probably a good chance to add nullability annotations to `LoginFields`, or even convert to Swift as it's basically a struct (even though it'd need to be a `class` for Obj-C compatibility)

aerych replied:
Macros added

koke commented:
There's one thing that bugs me about the new design and I think it's making the self hosted flow more confusing.

When you land in the new flow, there's a single field asking for your username or email, and a next button. This suggests a sort of wizard and grabs your attention so it's easy to miss there's an "Add self hosted" link at the bottom. Once you tap on the text field, it will be the last time you see that option:
![self-hosted-flow-fail](https://cloud.githubusercontent.com/assets/8739/14464809/c2397a94-00d0-11e6-83f0-02dc5a02efbb.png)

Compare this to the previous flow. You may not notice the link and enter your username and password, then get an error. Once you dismiss the error, you're back to the initial screen that you might scan more closely this time, noticing the self hosted option.

![self-hosted-flow-previous](https://cloud.githubusercontent.com/assets/8739/14464810/c25efb70-00d0-11e6-823e-581e15337bef.png)

koke commented on WordPress/Classes/ViewRelated/NUX/NUXAbstractViewController.swift:
```diff
@@ -0,0 +1,249 @@
+/// button and badge.
+/// It is assumed that NUX controllers will always be presented modally.
+///
+class NUXAbstractViewController : UIViewController
```
This looks like a common _chrome_ for all the NUX controllers. Did you consider making this a container instead of a base class?

aerych replied:
The original approach used parent/child view controllers.  It was a bit problematic in that we were maintaining more state than desired, wrangle a back button, and needing to manually manage the navigation history. We switching to a UINavigationController to simplify things and let it do the job it was intended for.

koke commented on WordPress/Classes/ViewRelated/NUX/NUXSubmitButton.swift:
```diff
@@ -0,0 +1,120 @@
+/// A stylized button used by NUX controllers. The button presents white text 
+/// surrounded by a white border.  It also can display a `UIActivityIndicatorView`.
+///
+@objc class NUXSubmitButton : UIButton
```
This looks similar in functionality to `PurchaseButton`/`RoundedButton` although using a regular `UIActivityIndicatorView` instead of `MRActivityIndicatorView`. Maybe we could combine them and have the subclasses deal with just the styling.

aerych replied:
I looked at this some.  They are functionally similar, but there doesn't look like a huge amount of overlap.  I'm not seeing where we'd gain much by extracting a common super class.

koke replied:
That's odd, I would have expected them to overlap more. We can leave it for now as the purchase is still a work in progress.

koke commented:
Some issues so far:

- The 1Password button from the email screen shows a list of unrelated logins
- From the self hosted screen with the keyboard up, if you drag to pop to the email controller there are some glitches on the 4s simulator. I don't see them on an iPad Air.
- In the Self Hosted login, the Next button on the (iPad) keyboard doesn't move to the next field for username/password
- When using an external keyboard, the forms still move as if there was a software keyboard on screen.
![screen shot 2016-04-12 at 18 51 40](https://cloud.githubusercontent.com/assets/8739/14468439/e6c6777c-00df-11e6-9067-e7e5b4929e1d.png)
![screen shot 2016-04-12 at 18 51 50](https://cloud.githubusercontent.com/assets/8739/14468440/e6ecca30-00df-11e6-9419-902b77409b1a.png)

koke commented on WordPress/Classes/ViewRelated/NUX/SigninWPComViewController.swift:
```diff
@@ -0,0 +1,279 @@
+    lazy var loginFacade: LoginFacade = {
+        let facade = LoginFacade()
+        facade.delegate = self
+        return facade
```
I don't think you're setting the delegate to `nil` on `deinit` so this happens:

1. Launch the app and enter your username
2. Enter whatever you want in the password field
3. Tap the Sign In button and as soon as possible tap Back
4. You get back to the Email screen and get an error. When you dismiss the error the screen is white.

![screen shot 2016-04-12 at 21 47 37](https://cloud.githubusercontent.com/assets/8739/14473484/a290fbb8-00f8-11e6-93c8-27d2cf38d492.png)

aerych replied:
Nice catch. I was able to reproduce this, thanks for the steps!  
I don't think it was the delegate.  The delegate has a weak reference so it shouldn't need to be set to nil. When testing it looked like the error vc was presented from the view controller that was being dismissed, which would leave things in a weird state for sure, but deinit wouldn't have been called at this point.  I've changed things so the error vc is instead presented from the nav controller. It would be better if LoginFacade had a halt/cancel/stop method we could call but that's out of scope here.

koke commented on WordPress/Classes/ViewRelated/NUX/SigninWPComSyncHandler.swift:
```diff
@@ -0,0 +1,88 @@
+protocol SigninWPComSyncHandler: class
+{
+    func configureLoading(loading: Bool)
+    func configureStatusMessage(message: String)
```
"configure" sounds odd when they just look like setters. I'd rather do `setStatusMessage(message: String)`. I'd prefer a property but not sure if that plays nice with ObjC `var statusMessage: String { get set }`

aerych replied:
I'd prefer to keep with the "setupMethodName" and "configureMethodName" pattern used within the signin vcs. Rather than making them look more like setters how about we make them look less like setters.

koke commented on WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift:
```diff
@@ -0,0 +1,249 @@
+        if !ReachabilityUtils.isInternetReachable() {
+            ReachabilityUtils.showAlertNoInternetConnection()
+            return
+        }
```
I recently learned from [a WWDC talk](https://developer.apple.com/videos/play/wwdc2012/706/) that Apple doesn't recommend using Reachability to *prevent* any connection. You should attempt the connection an if it fails, then use Reachability to determine when to retry.

aerych replied:
:+1:

aerych replied:
Reachability checks nuked.

aerych commented:
>  - The 1Password button from the email screen shows a list of unrelated logins

Fixed.

>  - From the self hosted screen with the keyboard up, if you drag to pop to the email controller there are some glitches on the 4s simulator. I don't see them on an iPad Air.

I tested on a 5 and didn't see any glitches (sorry, no 4s here). I'm inclined to call it a simulator glitch but let's test a bit more first.

>  - In the Self Hosted login, the Next button on the (iPad) keyboard doesn't move to the next field for username/password

Fixed

>  - When using an external keyboard, the forms still move as if there was a software keyboard on screen. 

Fixed.

aerych commented:
> I think it's making the self hosted flow more confusing.

I checked with @mattmiklic and we've added a button to the email/password from that will take the user to the self-hosted form.

aerych commented:
Thanks Koke! Ready for another round.

koke commented on WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift:
```diff
@@ -0,0 +1,275 @@
+    ///
+    class func isUsernameReserved(username: String) -> Bool {
+        let name = username.lowercaseString.trim()
+        return ["admin", "administrator", "root"].contains(name)
```
This won't make a difference specially with such a short list, so feel free to ignore, but a `Set` is more adequate than an `Array` here. From the `NSSet` documentation:

> You can use sets as an alternative to arrays when the order of elements isn’t important and performance in testing whether an object is contained in the set is a consideration—while arrays are ordered, testing for membership is slower than with sets.

The possible performance gains here would be irrelevant, so this is more about using the right type than any practical improvement.

aerych replied:
A fair point but since I'd likely still initialize the set with an array literal I'll leave it as is for more compact code.

koke commented on WordPress/Classes/System/WordPressAppDelegate.m:
```diff
@@ -434,6 +436,35 @@ - (void)application:(UIApplication *)application handleActionWithIdentifier:(NSS
+}
+
+
+/**
```
Can we move this inside `SignInHelpers` (or other helper class)?

aerych replied:
I wasn't sure which method this note was referring to so I made a best guess.  Let me know if I chose the wrong one.

koke commented:
> To test the final step, reopening the app passing the auth token, you'll need to copy the wordpress:// link from safari and change it to wpdebug://.

I'm not being able to do that, Safari just fails and the address bar is empty. Any tips?

![screen shot 2016-04-14 at 10 27 32](https://cloud.githubusercontent.com/assets/8739/14521856/ccc352bc-022b-11e6-8e1b-7b96f127d2c5.png)

koke commented on WordPress/Classes/System/WordPressAppDelegate.m:
```diff
@@ -518,8 +549,63 @@ - (void)showWelcomeScreenIfNeededAnimated:(BOOL)animated
+        UINavigationController *navigationController = [[NUXNavigationController alloc] initWithRootViewController:controller];
+        UIViewController *rootViewController = self.window.rootViewController;
+        if (rootViewController.presentedViewController)  {
+            [rootViewController.presentedViewController dismissViewControllerAnimated:NO completion:^{
```
Why do we need dismissing here? I fear there's a chance this could dismiss an unsaved post or comment reply. I haven't been able to test the magic link flow, and the new sign in only seems to be enabled for now for the first time setup. But I would expect this could make its way into all the others sign in flows, so my worry is that the following could happen:

1. You have a working self hosted blog in the app already
2. Try to log in to wp.com using the email link.
3. Email is a bit slow today so you give up for now and go back to replying to comments on your self hosted.
4. A notification pops up with new mail
5. Go to mail, tap the link, this is called, and your half written comment is lost

Another potential problematic flow:

1. You install the app and try to log in to wp.com using the magic link
2. Email is not arriving, so you decide to use the password instead
3. You start writing a post
4. The email finally arrives, and for some reason you tap the link
5. This is called, the editor is dismiss (I'd hope the content isn't lost though)
6. Not sure what else happens, but the rest looks safe

aerych replied:
> Why do we need dismissing here?

If a UIViewController already has a presentedViewController it will not present another before the first is dismissed.   Its possible when the auth link is opened for the app is resumed vs launched in which case one of the other signin controllers would still be presented. 

Good thinking about the other uses cases. I can tweak the logic here.

koke commented on WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift:
```diff
@@ -0,0 +1,248 @@
+
+    /// Configures the appearance and state of the submit button.
+    ///
+    func configureSubmitButton(animating: Bool) {
```
I'd make `animating` a named parameter (also in the other VCs that have this). When you see something like `configureSubmitButton(false)` it's really hard to know what the parameter means. `configureSubmitButton(animating: false)` is much more obvious.

koke commented on WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift:
```diff
@@ -0,0 +1,275 @@
+
+        path = wpLogin.stringByReplacingMatchesInString(path, options: .ReportCompletion, range: NSRange(location: 0, length: path.characters.count), withTemplate: "")
+        path = wpadmin.stringByReplacingMatchesInString(path, options: .ReportCompletion, range: NSRange(location: 0, length: path.characters.count), withTemplate: "")
+        path = trailingSlash.stringByReplacingMatchesInString(path, options: .ReportCompletion, range: NSRange(location: 0, length: path.characters.count), withTemplate: "")
```
Small (optional) suggestion, make a `String.dropSuffix(_)` helper

aerych replied:
Adding helper if for no other reason than to have the utility going forward, but if you want to do a PR for that super cool NSURL extension you shared I'd be game to use that instead :)

koke replied:
:+1: We can leave that for another PR, since you mentioned this was just porting the existing Obj-C to Swift.

aerych commented:
> Safari just fails and the address bar is empty. Any tips?

Try opening the email and tapping the link in the desktop Safari browser.  You should be able to copy and edit the `wordpress://` from the error prompt that appears.

aerych commented:
I *think* that's handled everything for this round.  Tag @koke :)

koke commented on WordPress/Classes/Extensions/String+Helpers.swift:
```diff
@@ -10,8 +10,24 @@ extension String {
+    /// - Parameters: 
+    ///     - suffix: The suffxing to remove.  This can be a regular expression but should not include an ending `$`.
+    ///
+    func trimSuffix(suffix: String) -> String {
```
The documentation does say that it expects a regex, but it can be easy to miss. I'd use a named parameter so it's obvious `string.trimSuffix(regexp: "wp-admin/?")`. Otherwise I'd expect it to do an exact string match

frosty commented on WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift:
```diff
@@ -0,0 +1,248 @@
+            NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType
+        ]
+
+        let styledString = "<style>body {font-family: sans-serif; font-size:14px; color: #ffffff; text-align:center;}</style>" + string
```
I've just been doing this same thing for a footer in Plans that has some highlighting. It looks like a font of `sans-serif` uses Helvetica instead of San Francisco. You'll need to use a value of `-apple-system` here if you want SF.

aerych replied:
Ah!  Thanks for the tip :)

koke commented:
This is a small detail, and not sure if new, but:

1. Launch app
2. Enter wrong username, tap next
3. Enter password, tap sign in
4. Get an error
5. Tap on 1Password
6. Realize you were using the wrong username 
7. Tap on the right credentials
8. The app starts trying to sign in, but the username in the text field is still what you typed

If the 1Password credentials are good, it'll sign in correctly, but I found it a bit confusing, specially if the connection is a bit slow.

koke commented:
Another fun bug:

1. Launch the app
2. Enter username, tap next
3. Enter password, tap sign in and immediately tap back
4. Enter another account username, tap next
5. Enter another account password, tap sign in

You're logged in as the first account. You can find the sites for both accounts in My Sites.

aerych commented:
> The app starts trying to sign in, but the username in the text field is still what you typed

Good catch. 

> You're logged in as the first account. You can find the sites for both accounts in My Sites. 

The horror... the horror...

kurzee commented:
>You're logged in as the first account. You can find the sites for both accounts in My Sites.

That's similar to this issue I've been looking at https://github.com/wordpress-mobile/WordPress-iOS/issues/5042. One problem here is that `AccountService` doesn't clean up the previous default account, so edge cases like this one end up creating two default accounts.

aerych commented:
@koke, the bugs should be addressed.  Final check?

koke commented:
:shipit: ❗ Really nice work here.

aerych commented:
Thanks @koke !! :)
